### PR TITLE
Defragmentation Memos

### PR DESCRIPTION
**Pull request by dolanbernard**
* Added DefragmentationMemoBuilder
* Added DefragmentationMemo

### Motivation

This PR implements changes outlined in [MCIP #61](https://github.com/mobilecoinfoundation/mcips/pull/61).

[MCIP #61](https://github.com/mobilecoinfoundation/mcips/pull/61) proposes a new memo type: the defragmentation memo. This memo type is used to denote defragmentation transactions. It also includes an ID field that can be used to easily group together multiple rounds of one defragmentation session together for the purpose of reporting total paid in fees and total amount defragmented. The full implementation details and rationale for defragmentation memos can be found [here](https://github.com/dolanbernard/mcips/blob/defrag-memo/text/0061-defrag-memos.md).

### CODE REVIEW

File: transaction/extra/src/memo/defragmentation.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+/// An error that can occur when configuring a destination memo
+#[derive(Display, Debug)]
+pub enum DefragmentationMemoError {
+    /// The fee amount is too large to be represented in the DefragmentationMemo
+    FeeTooLarge,
+}
+
+impl_memo_type_conversions! { DefragmentationMemo }
```
Comment (nick-mobilecoin) [inline transaction/extra/src/memo/defragmentation.rs:144]: :wrench: Not sure why the destination memo test is in `memo/mod.rs` but suggest putting the test for the defragmentation memo in here.

There are a few changes that may not be agreeable so take anything you like:
- remove name prefix of `test_` that's generally used for discovery for xunit frameworks while the `#[test]` macro in rust provides the discovery
- Broke the tests out to separate test functions. We have quite a few large tests so would like to see smaller more focused ones. The test stops on the first failure so if multiple things would fail you only get the first one.
- Explicit comparison against the expected error, this requires adding `PartialEq` to the Error derives
- Dropped `u64` from the literals, fetl kind of noisy, but explicit is nice :shrug: 
- Used a constant for the boundary tests to max fee
- Removed the bitwise not on the id usage, it's felt like it might be easy to miss.
```suggestion
impl_memo_type_conversions! { DefragmentationMemo }

#[cfg(test)]
mod test {
    use super::*;

    const MAX_FEE: u64 = 0x00FFFFFF_FFFFFFFF;

    #[test]
    fn getters_and_setters() {
        let mut memo = DefragmentationMemo::new(48237, 9001, 3405697037).unwrap();

        assert_eq!(48237, memo.get_fee());
        assert_eq!(9001, memo.get_total_outlay());
        assert_eq!(3405697037, memo.get_defrag_id());

        memo.set_fee(73284).unwrap();
        memo.set_total_outlay(8999);
        memo.set_defrag_id(10);

        assert_eq!(73284, memo.get_fee());
        assert_eq!(8999, memo.get_total_outlay());
        assert_eq!(10, memo.get_defrag_id());
    }

    #[test]
    fn max_fee_succeeds() {
        let memo = DefragmentationMemo::new(MAX_FEE, 0, 0).unwrap();
        assert_eq!(MAX_FEE, memo.get_fee());
    }

    #[test]
    fn exceeding_max_fee_fails() {
        let mut memo = DefragmentationMemo::new(0, 0u64, 0u64).unwrap();
        assert_eq!(
            memo.set_fee(MAX_FEE + 1),
            Err(DefragmentationMemoError::FeeTooLarge)
        );
    }

    #[test]
    fn new_propagates_error() {
        assert_eq!(
            DefragmentationMemo::new(MAX_FEE + 1, 0u64, 0u64),
            Err(DefragmentationMemoError::FeeTooLarge)
        );
    }

    #[test]
    fn serialization() {
        let memo = DefragmentationMemo::new(2525, 36, 80).unwrap();

        let memo_bytes: [u8; DefragmentationMemo::MEMO_DATA_LEN] = memo.clone().into();
        let deserialized_memo = DefragmentationMemo::from(&memo_bytes);

        assert_eq!(memo, deserialized_memo);
    }
}

```

File: transaction/extra/src/memo/mod.rs [hunk 2]
```diff
@@ -415,4 +419,30 @@ mod tests {
         assert_eq!(memo.get_fee(), 17u64);
         assert_eq!(memo.get_num_recipients(), 4);
     }
+
+    #[test]
+    fn test_defragmentation_memo() {
```
Comment (nick-mobilecoin) [inline transaction/extra/src/memo/mod.rs:424]: :+1: :+1: Really like these tests:
- Testing the boundary condition of max fee
- Testing the getters and setters, even though their straightforward the tests aren't that costly to write.
- Testing the error out of `new()`
- Testing serialzing

:+1:

File: transaction/extra/src/memo/defragmentation.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    fn from(src: DefragmentationMemo) -> [u8; DefragmentationMemo::MEMO_DATA_LEN] {
+        let mut memo_data = [0u8; DefragmentationMemo::MEMO_DATA_LEN];
+        memo_data[..7].copy_from_slice(&src.fee.to_be_bytes()[1..]);
+        memo_data[7..15].copy_from_slice(&src.total_outlay.to_be_bytes());
+        memo_data[15..23].copy_from_slice(&src.defrag_id.to_be_bytes());
+        memo_data
+    }
+}
+
+/// An error that can occur when configuring a destination memo
```
Comment (nick-mobilecoin) [inline transaction/extra/src/memo/defragmentation.rs:137]: :upside_down_face: 
```suggestion
/// An error that can occur when configuring a defragmentation memo
```

File: transaction/builder/src/transaction_builder.rs [hunk 4]
```diff
@@ -4129,4 +4130,361 @@ pub mod transaction_builder_tests {
             );
         }
     }
+
+    #[test]
+    // Transaction builder with Burn Redemption memo builder
+    fn test_transaction_builder_defragmentation_memos() {
```
Comment (nick-mobilecoin) [inline transaction/builder/src/transaction_builder.rs:4136]: :wrench: I know the other test functions do it so...
but I don't think this comment provides any value over the function name especially since it's wrong.
```suggestion
    fn transaction_builder_defragmentation_memos() {
```

File: transaction/builder/src/transaction_builder.rs [hunk 5]
```diff
@@ -4129,4 +4130,361 @@ pub mod transaction_builder_tests {
...
+            let mut transaction_builder = TransactionBuilder::new(
+                block_version,
+                Amount::new(10, token_id),
+                fog_resolver.clone(),
+                memo_builder,
+            )
+            .unwrap();
+
+            transaction_builder.set_fee(3).unwrap();
+
```
Comment (nick-mobilecoin) [inline transaction/builder/src/transaction_builder.rs:4281]: :wrench: Not seeing a need for the fee change here
```suggestion
```

File: transaction/builder/src/transaction_builder.rs [hunk 6]
```diff
@@ -4129,4 +4130,361 @@ pub mod transaction_builder_tests {
...
```
Comment (nick-mobilecoin) [inline transaction/builder/src/transaction_builder.rs:4321]: :wrench: Similar to above fee doesn't seem necessary
```suggestion
```

File: transaction/builder/src/transaction_builder.rs [hunk 7]
```diff
@@ -4129,4 +4130,361 @@ pub mod transaction_builder_tests {
...
```
Comment (nick-mobilecoin) [inline transaction/builder/src/transaction_builder.rs:4412]: :wrench: similar to above, fee seems unecessary
```suggestion
```

File: fog/sample-paykit/src/cached_tx_data/memo_handler.rs [hunk 8]
```diff
@@ -149,6 +149,13 @@ impl MemoHandler {
                     Err(MemoHandlerError::FailedSubaddressValidation)
                 }
             }
+            MemoType::Defragmentation(_) => {
+                if subaddress_matches_tx_out(account_key, CHANGE_SUBADDRESS_INDEX, tx_out)? {
+                    Ok(Some(memo_type))
+                } else {
+                    Err(MemoHandlerError::FailedSubaddressValidation)
+                }
+            }
```
[the remark this answers is not in the document]
Comment (NotGyro) [reply fog/sample-paykit/src/cached_tx_data/memo_handler.rs:158]: The second option seems more reasonable to me but I don't have a strong preference either way, it might be a nice thing to correct later but I'd be fine with the PR getting merged with this code as-is.

File: transaction/extra/src/memo/defragmentation.rs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+/// transaction). The memo written to the main TxOut will have the fee
+/// and total outlay recorded in their respective fields. The change
+/// TxOut will have a memo with 0 fee and 0 outlay. This makes
+/// calculating the total fee and total outlays easier, as the change
+/// TxOut does not need to be manually detected and ignored. Both the
+/// main and change memos will receive the same value for the
+/// defragmentation ID.
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct DefragmentationMemo {
+    /// The fee paid to perform the defragmentation transaction (picoMOB)
```
[the remark this answers is not in the document]
Comment (NotGyro) [reply transaction/extra/src/memo/defragmentation.rs:33]: Could be `(in the base units of the token ID - for instance, picoMOB in MOB's case, and micro-eUSD in eUSD's case)`

**Review by NotGyro — APPROVED** [review]
This looks good to go to me, so long as the issues @nick-mobilecoin has already identified are resolved.

File: transaction/builder/src/transaction_builder.rs (hunk 6 above)
Comment (dolanbernard) [reply transaction/builder/src/transaction_builder.rs:4318]: This one is actually necessary. Below, in the first set of asserts, the fee in the main defrag output should match the value set here. The second set of asserts checks that even though the fee was non-zero, the "change" output is built with a fee of zero. Rationale for this behavior is discussed and specified in [MCIP 61](https://github.com/mobilecoinfoundation/mcips/pull/61).

File: transaction/builder/src/transaction_builder.rs (hunk 7 above)
Comment (dolanbernard) [reply transaction/builder/src/transaction_builder.rs:4409]: Same as above.

**Review by nick-mobilecoin — APPROVED** [review]
LGTM Thanks!
